### PR TITLE
CPack Debian PPA Build

### PR DESCRIPTION
**Pull request by daschuer**
This adds a CPack target to Upload to https://launchpad.net/~mixxx/+archive/ubuntu/mixxxbetas

By 
```
cpack -G External -D DEB_UPLOAD_PPA=ppa:mixxx/mixxxbetas
```

It used debuild for packaging. Which is well optimised with cmake the required rules file is almost empty. 

The folder structure is
- packaging
--- CPack Files 
--- debian 
----- debian Files   

The common CPack variables are re-used.  

@uklotzde: 
This moves the udev file to the correct location, that you can remove the move hack from the rpmfusion script.
lintian was complaining about the redundant font installation. So I have removed them from the linux install() and added them as package dependency. The same need to be done for rpmfusion as well. 

How about creating a: 
packaging/rpmfusion 
folder as well and maintain the spec file by CPack as well? This would allow us to keep everything in sync.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1286,13 +1286,19 @@ if(UNIX AND NOT APPLE)
...
-  )
+  option(INSTALL_USER_UDEV_RULES "Install user udev rule file for USB HID and Bulk controllers" ON)
+  if (INSTALL_USER_UDEV_RULES)
+    install(
+      FILES
+        "${CMAKE_CURRENT_SOURCE_DIR}/res/linux/mixxx-usb-uaccess.rules"
+      DESTINATION
+        "${CMAKE_INSTALL_SYSCONFDIR}/udev/rules.d"
+    )
+    message(WARNING "Installing udev rules to ${CMAKE_INSTALL_SYSCONFDIR}/udev/rules.d, "
```
Comment (Be-ing) [inline CMakeLists.txt:1297]: https://github.com/daschuer/mixxx/pull/62

File: CMakeLists.txt [hunk 2]
```diff
@@ -2510,7 +2531,8 @@ endif()
 # Packaging
 set(CPACK_PACKAGE_VENDOR "Mixxx Project")
 set(CPACK_PACKAGE_CONTACT "RJ Skerry-Ryan <rryan@mixxx.org>")
-set(CPACK_PACKAGE_DESCRIPTION_FILE "${CMAKE_CURRENT_SOURCE_DIR}/cmake/cpack_package_description.txt")
+set(CPACK_PACKAGE_DESCRIPTION_SUMMARY "Digital Disc Jockey Application")
```
Comment (Be-ing) [inline CMakeLists.txt:2534]:
```suggestion
set(CPACK_PACKAGE_DESCRIPTION_SUMMARY "Digital DJ Application")
```
I think it is more likely a user would search their package manager for "DJ" than "Disc Jockey".

File: CMakeLists.txt [hunk 3]
```diff
@@ -2522,12 +2544,42 @@ set(CPACK_RESOURCE_FILE_README "${CMAKE_CURRENT_SOURCE_DIR}/README")
 set(CPACK_STRIP_FILES ON)
 set(CPACK_CREATE_DESKTOP_LINKS "mixxx")
 
+set(CPACK_SOURCE_IGNORE_FILES  "\\\\.#;/#;.*~;\\\\.o$")
+list(APPEND CPACK_SOURCE_IGNORE_FILES "/\\\\.git/")
+list(APPEND CPACK_SOURCE_IGNORE_FILES "/\\\\.github/")
+list(APPEND CPACK_SOURCE_IGNORE_FILES "/cache/")
+list(APPEND CPACK_SOURCE_IGNORE_FILES "/.*_build/")
+list(APPEND CPACK_SOURCE_IGNORE_FILES "/\\\\.sconf_temp/")
```
Comment (Be-ing) [inline CMakeLists.txt:2552]: I think you can remove this line at this point.

File: CHANGELOG.md [hunk 4]
```diff
@@ -128,8 +128,6 @@
 
 ## [2.2.2](https://launchpad.net/mixxx/+milestone/2.2.2) (2019-08-10)
 
-==== 2.2.2 2019-08-10 ====
-
```
Comment (Holzhaus) [inline CHANGELOG.md:132]: Unrelated?

File: packaging/CPackDebUploadPPA.cmake [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  message(FATAL_ERROR "debchange not found, required for cpack -G External -D DEB_UPLOAD_PPA=true" )
+endif()
+
+find_program(CPACK_DEBIAN_MARKDOWN markdown)
+if(NOT CPACK_DEBIAN_MARKDOWN)
+  message(FATAL_ERROR "debchange not found, required for cpack -G External -D DEB_UPLOAD_PPA=true" )
+endif()
+
+
+message( NOTICE "Creating mixxx_${CPACK_DEBIAN_UPSTREAM_VERSION}.orig.tar.gz" )
```
Comment (Holzhaus) [inline packaging/CPackDebUploadPPA.cmake:26]:
```suggestion
message(NOTICE "Creating mixxx_${CPACK_DEBIAN_UPSTREAM_VERSION}.orig.tar.gz" )
```

File: packaging/CPackConfig.cmake [hunk 6]
```diff
@@ -0,0 +1,16 @@
+# unlike CMakeLists.txt this file is include at cpack time, once per generator after CPack has set CPACK_GENERATOR
+# to the actual generator being used. It allows per-generator setting of CPACK_* variables at cpack time.
+
+if (CPACK_GENERATOR MATCHES "DEB")
+  set( CPACK_INSTALL_SCRIPT ${CPACK_DEBIAN_INSTALL_SCRIPT} )
```
Comment (Holzhaus) [inline packaging/CPackConfig.cmake:5]:
```suggestion
  set(CPACK_INSTALL_SCRIPT ${CPACK_DEBIAN_INSTALL_SCRIPT} )
```

File: packaging/CPackDebUploadPPA.cmake [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+endif()
+
+find_program(CPACK_DEBIAN_DEBCHANGE debchange)
+if(NOT CPACK_DEBIAN_DEBCHANGE)
+  message(FATAL_ERROR "debchange not found, required for cpack -G External -D DEB_UPLOAD_PPA=true" )
+endif()
+
+find_program(CPACK_DEBIAN_MARKDOWN markdown)
+if(NOT CPACK_DEBIAN_MARKDOWN)
+  message(FATAL_ERROR "debchange not found, required for cpack -G External -D DEB_UPLOAD_PPA=true" )
```
Comment (Holzhaus) [inline packaging/CPackDebUploadPPA.cmake:22]:
```suggestion
  message(FATAL_ERROR "markdown not found, required for cpack -G External -D DEB_UPLOAD_PPA=true" )
```

**Review by Holzhaus — COMMENTED** [review]
Can you split uploading from packaging? That way we can test the cpack scripts in PRs without uploading them to the PPA. Something like this:

```udiff
diff --git a/packaging/CPackConfig.cmake b/packaging/CPackConfig.cmake
index 2514ac5815..55ccbc877c 100644
--- a/packaging/CPackConfig.cmake
+++ b/packaging/CPackConfig.cmake
@@ -6,7 +6,7 @@ if (CPACK_GENERATOR MATCHES "DEB")
 endif()

 if (CPACK_GENERATOR MATCHES "External")
-  if (DEB_UPLOAD_PPA)
+  if (DEB_SOURCEPKG OR DEB_UPLOAD_PPA)
     set(CPACK_EXTERNAL_ENABLE_STAGING true)
     set(CPACK_INSTALLED_DIRECTORIES "${CPACK_DEBIAN_SOURCE_DIR};/")
     set(CPACK_IGNORE_FILES "${CPACK_SOURCE_IGNORE_FILES}")
diff --git a/packaging/CPackDebUploadPPA.cmake b/packaging/CPackDebUploadPPA.cmake
index c73da2b4c5..0e07c836e1 100644
--- a/packaging/CPackDebUploadPPA.cmake
+++ b/packaging/CPackDebUploadPPA.cmake
@@ -7,9 +7,11 @@ if(NOT CPACK_DEBIAN_DEBUILD)
   message(FATAL_ERROR "debuild not found, required for cpack -G External -D DEB_UPLOAD_PPA=true" )
 endif()

-find_program(CPACK_DEBIAN_DPUT dput)
-if(NOT CPACK_DEBIAN_DPUT )
-  message(FATAL_ERROR "dput not found, required for cpack -G External -D DEB_UPLOAD_PPA=true" )
+if(DEB_UPLOAD_PPA)
+  find_program(CPACK_DEBIAN_DPUT dput)
+  if(NOT CPACK_DEBIAN_DPUT)
+    message(FATAL_ERROR "dput not found, required for cpack -G External -D DEB_UPLOAD_PPA=true" )
+  endif()
 endif()

 find_program(CPACK_DEBIAN_DEBCHANGE debchange)
@@ -63,8 +65,10 @@ foreach(RELEASE ${CPACK_DEBIAN_DISTRIBUTION_RELEASES})
        WORKING_DIRECTORY ${CPACK_TOPLEVEL_DIRECTORY}/${CPACK_PACKAGE_FILE_NAME}
        RESULT_VARIABLE CPACK_DEBIAN_DEBUILD_RET)
   if(NOT CPACK_DEBIAN_DEBUILD_RET EQUAL "0")
-    message(FATAL_ERROR "${CPACK_DEBIAN_DEBUILD} retuned exit code ${CPACK_DEBIAN_DEBUILD_RET}")
+    message(FATAL_ERROR "${CPACK_DEBIAN_DEBUILD} returned exit code ${CPACK_DEBIAN_DEBUILD_RET}")
+  endif()
+  if(DEB_UPLOAD_PPA)
+    execute_process(COMMAND ${CPACK_DEBIAN_DPUT} ${DEB_UPLOAD_PPA} "mixxx_${CPACK_DEBIAN_UPSTREAM_VERSION}-${CPACK_DEBIAN_DEBIAN_VERSION}~${CPACK_DEBIAN_DEBIAN_VERSION_EXTRA}~${RELEASE}_source.changes"
+         WORKING_DIRECTORY ${CPACK_TOPLEVEL_DIRECTORY})
   endif()
-  execute_process(COMMAND ${CPACK_DEBIAN_DPUT} ${DEB_UPLOAD_PPA} "mixxx_${CPACK_DEBIAN_UPSTREAM_VERSION}-${CPACK_DEBIAN_DEBIAN_VERSION}~${CPACK_DEBIAN_DEBIAN_VERSION_EXTRA}~${RELEASE}_source.changes"
-       WORKING_DIRECTORY ${CPACK_TOPLEVEL_DIRECTORY})
```

**Comment by Holzhaus** [comment]
Btw, I installed the `devscripts` and `discount` packages on my Arch Linux install, but building the source package fails:

```
CPack: Create package using External
CPack: Install projects
CPack: - Install directory: /home/jan/Projects/mixxx
CPack: Create package
Creating mixxx_2.3.0~beta.orig.tar.gz
Creating debian folder
CMake Error at /home/jan/Projects/mixxx/packaging/CPackDebUploadPPA.cmake:49 (file):
  file COPY cannot copy file
  "/home/jan/Projects/mixxx/cmake_build/_CPack_Packages/Linux/External/Mixxx-debian_ppa-r7856+-1fdca9e44e/res/linux/mixxx-usb-uaccess.rules"
  to
  "/home/jan/Projects/mixxx/cmake_build/_CPack_Packages/Linux/External/Mixxx-debian_ppa-r7856+-1fdca9e44e/debian/control/mixxx-usb-uaccess.rules":
  Not a directory.


CMake Error at /home/jan/Projects/mixxx/packaging/CPackDebUploadPPA.cmake:51 (file):
  file RENAME failed to rename

    /home/jan/Projects/mixxx/cmake_build/_CPack_Packages/Linux/External/Mixxx-debian_ppa-r7856+-1fdca9e44e/debian/control/mixxx-usb-uaccess.rules

  to

    /home/jan/Projects/mixxx/cmake_build/_CPack_Packages/Linux/External/Mixxx-debian_ppa-r7856+-1fdca9e44e/debian/control/mixxx.mixxx-usb-uaccess.rules

  because: Not a directory



libdistro-info-perl is not installed, Debian release names are not known.
libdistro-info-perl is not installed, Ubuntu release names are not known.
 dpkg-buildpackage -us -uc -ui -S -sa
dpkg-buildpackage: info: source package mixxx
dpkg-buildpackage: info: source version 2.3.0~beta-0ubuntu1~debianppa~git7856+~bionic
dpkg-buildpackage: info: source distribution bionic
dpkg-buildpackage: info: source changed by RJ Skerry-Ryan <rryan@mixxx.org>
 dpkg-source --before-build .
dpkg-checkbuilddeps: error: Unmet build dependencies: debhelper (>= 11) pkg-config docbook-to-man markdown libglu1-mesa-dev qt5-default qtscript5-dev libqt5opengl5-dev libqt5svg5-dev libqt5sql5-sqlite libqt5x11extras5-dev cmake libjack-dev portaudio19-dev libid3tag0-dev libmad0-dev libogg-dev libsndfile1-dev libasound2-dev libvorbis-dev libfaad-dev libportmidi-dev libtag1-dev libshout-dev libssl-dev libprotobuf-dev protobuf-compiler libusb-1.0-0-dev libchromaprint-dev librubberband-dev libopusfile-dev libsqlite3-dev libsoundtouch-dev libhidapi-dev libupower-glib-dev liblilv-dev libmodplug-dev libmp3lame-dev libebur128-dev xvfb
dpkg-buildpackage: warning: build dependencies/conflicts unsatisfied; aborting
dpkg-buildpackage: warning: (Use -d flag to override.)
debuild: fatal error at line 1182:
dpkg-buildpackage -us -uc -ui -S -sa failed
CMake Error at /home/jan/Projects/mixxx/packaging/CPackDebUploadPPA.cmake:68 (message):
  /usr/bin/debuild returned exit code 29


CPack Error: Problem compressing the directory
CPack Error: Error when generating package: mixxx
```

File: packaging/CPackConfig.cmake (hunk 6 above)
Comment (Be-ing) [reply packaging/CPackConfig.cmake:5]: `  set(CPACK_INSTALL_SCRIPT ${CPACK_DEBIAN_INSTALL_SCRIPT})`

File: CHANGELOG.md (hunk 4 above)
Comment (daschuer) [reply CHANGELOG.md:132]: Yes. Is this a problem?

File: CMakeLists.txt (hunk 3 above)
Comment (daschuer) [reply CMakeLists.txt:2552]: The file is most likely still there on a users machine.

File: packaging/CPackDebUploadPPA.cmake [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+# cpack -G External -D DEB_UPLOAD_PPA=true
+
+find_program(CPACK_DEBIAN_DEBUILD debuild)
+if(NOT CPACK_DEBIAN_DEBUILD)
+  message(FATAL_ERROR "debuild not found, required for cpack -G External -D DEB_UPLOAD_PPA=true" )
+endif()
+
+if(DEB_UPLOAD_PPA)
+  find_program(CPACK_DEBIAN_DPUT dput)
+  if(NOT CPACK_DEBIAN_DPUT )
```
Comment (Be-ing) [inline packaging/CPackDebUploadPPA.cmake:12]:
```suggestion
  if(NOT CPACK_DEBIAN_DPUT)
```

**Comment by daschuer** [comment]
Successfully tested here: https://launchpad.net/~daschuer/+archive/ubuntu/mixxx/+packages

**Comment by Be-ing** [comment]
@Holzhaus ready to merge?

File: packaging/CPackDebUploadPPA.cmake [hunk 9]
```diff
@@ -24,6 +24,11 @@ if(NOT CPACK_DEBIAN_MARKDOWN)
   message(FATAL_ERROR "markdown not found, required for cpack -G External -D DEB_UPLOAD_PPA=true")
 endif()
 
+# hack from sconscript
+#if (DEB_UPLOAD_PPA MATCHES "mixxxbetas")
```
Comment (Be-ing) [inline packaging/CPackDebUploadPPA.cmake:28]: Why is the commented code still there?

**Comment by Holzhaus** [comment]
https://github.com/daschuer/mixxx/pull/63 please check if this is correct, it looks like ig just generates a json file?!

File: CMakeLists.txt [hunk 10]
```diff
@@ -1294,10 +1294,17 @@ if(UNIX AND NOT APPLE)
       DESTINATION
         "${CMAKE_INSTALL_SYSCONFDIR}/udev/rules.d"
     )
-    message(WARNING "Installing udev rules to ${CMAKE_INSTALL_SYSCONFDIR}/udev/rules.d, "
-        "which is reserved for user-installed files. The correct directory for "
-        "system rules is /lib/udev/rules.d. with a appropiated priority prefix. "
-        "Adjust your package script accordingly and set -DINSTALL_USER_UDEV_RULES=OFF")
+    message(WARNING
+"The udev rule file for USB HID and Bulk controller permissions will be installed to "
```
Comment (Be-ing) [inline CMakeLists.txt:1298]: There are multiple formatting issues in this. Please reset this commit and merge my PR instead. Or just reset this commit and I'll open a followup PR to 2.3.

**Comment by daschuer** [comment]
It is failing now because of missing PGP key. I will disable the GitHub build now. Jenkins will than do the job after merging.

**Comment by daschuer** [comment]
Done.

**Comment by daschuer** [comment]
@uklotzde: Did you consider my initial comments?

**Comment by uklotzde** [comment]
> @uklotzde: Did you consider my initial comments?

The .spec file may differ for different Fedora releases (separate branch for each release) and also contains Fedora-specific macros. This won't fit into the Mixxx source repo. The Mixxx source repo should be agnostic to any distributions and only provide the building blocks.

**Comment by uklotzde** [comment]
Strictly speaking the source repo should not contain any packaging code at all. This should better be maintained in a separate repo. Unrelated to this PR. I don't mind keeping the Ubuntu packaging for historical reasons as long as the shared artifacts remain distribution agnostic.

**Comment by uklotzde** [comment]
Packaging requires a different versioning scheme. That's the main reason why it doesn't fit into the source repo.

File: res/linux/mixxx.appdata.xml [hunk 11]
```diff
@@ -36,6 +36,25 @@
       <li>Skinnable interface with several skins bundled</li>
     </ul>
   </description>
+  <provides>
+    <modalias>usb:v22F0p0008d*</modalias>
```
Comment (uklotzde) [inline res/linux/mixxx.appdata.xml:40]: If these IDs should match those in mixxx-usb-uaccess.rules we should add a comment in **both** files!

**Comment by daschuer** [comment]
> The .spec file may differ for different Fedora releases (separate branch for each release) and also contains Fedora-specific macros. This won't fit into the Mixxx source repo. The Mixxx source repo should be agnostic to any distributions and only provide the building blocks.

This is the same issue for Ubuntu as well.

Here it works like that we have the templates for all derivatives in the deployment folder.
Here in the subfolder "debian". Something like a rpmforge/mixxx.spec.in
A CPack script is then used to prepare the version specific output. A fedora_17/mixxx.spec

There is one big advantage to backport package script templates into our source tree. They can be automated updated when something special is to do, like an updated description. 

For instance in the package description now in a single place.

**Comment by daschuer** [comment]
> Packaging requires a different versioning scheme. That's the main reason why it doesn't fit into the source repo.

Yes, that's the same for Ubuntu as well.
The generated package has its own version.

In the Fedora case you can than upload the configured files to the rpmfusion git. The workflow remains, just one manual step is automated.

**Comment by Be-ing** [comment]
Couldn't you generate the source DEB on GH Actions but not upload it to the PPA until @rryan adds the GPG key?

File: res/linux/mixxx.appdata.xml (hunk 11 above)
Comment (daschuer) [reply res/linux/mixxx.appdata.xml:40]: mixxx-usb-uaccess.rules is still there.
It schould only not be installed to etc first. You mixxx.spec relies on that.
I have added a flag to control this.

File: CMakeLists.txt (hunk 10 above)
Comment (Be-ing) [reply CMakeLists.txt]: Please rebase to remove this commit.

**Comment by daschuer** [comment]
> Couldn't you generate the source DEB on GH Actions but not upload it to the PPA until @rryan adds the GPG key?

That would conflict with the signing required on Jenkins. Update our beta users with PPA builds is more important.  We will have the source packaged on the PPA once this is merged anyway. 

If Jenkins dies before RJ has uploaded his certificate, we can use mine for both.

**Comment by Be-ing** [comment]
What is the conflict? I don't understand.

**Comment by daschuer** [comment]
The source package is signed as well.
This needs to be the same authority that signs the Deb.

File: res/linux/mixxx.appdata.xml (hunk 11 above)
Comment (uklotzde) [reply res/linux/mixxx.appdata.xml:41]: Cool, thank you :+1:

**Comment by Be-ing** [comment]
Why must the source package be signed? That should be optional if a signing key is available. It should be easy to generate a source DEB locally without bothering to set up code signing.

**Comment by daschuer** [comment]
It would be possible with a bit of extra work.
However cpack can already create unsigned source packages. So there is no point for doing this.

**Comment by daschuer** [comment]
What is open for merging this?

**Comment by Be-ing** [comment]
Please rebase to remove 37099e0 then I will open a followup PR to 2.3.

**Comment by Holzhaus** [comment]
> It would be possible with a bit of extra work.
> However cpack can already create unsigned source packages. So there is no point for doing this. 

If CPack can already create unsigned Debian source packages, can't we just use that and sign it afterwards and upload that to the PPA? Are all of these changes just for signing and uploading?

It would be good to make it possible to build unsigned debian source packages, because then we can test it on CI builds for PRs. Secrets are not available for PR builds, because that would make it possible to exfiltrate them, so making signed sources packages for PRs does not work.

**Comment by daschuer** [comment]
> Please rebase to remove 37099e0 then I will open a followup PR to 2.3.

What are you going to do? 

Feel free to do a new PR on top of this. If it turns out that a rebase is necessary, I can do it after merging your PR.

**Comment by daschuer** [comment]
> If CPack can already create unsigned Debian source packages, can't we just use that and sign it afterwards and upload that to the PPA?

This PR basically does that. 

This PR works:
cpack->debuild->debhelper->cmake.

We may think of calling the signing via debhelper directly. However, the debuild tool chain does this already very well. It knows in addition special things of the differen Ubuntu releases and includes many sanity checks via lintian. 

Also it is almost perfectly integrated with cmake for the final build step.

Are all of these changes just for signing and uploading?

Not only. It generates also the package metadata from the CMakeList.txt. That keeps all metadata for Debian, wix and mac package in a single place.

Thats relates to the open discussion to do the same with rpmforge. 

> It would be good to make it possible to build unsigned debian source packages, because then we can test it on CI builds for PRs.

> Secrets are not available for PR builds, because that would make it possible to exfiltrate them, so making signed sources packages for PRs does not work. 

Yes that is not a problem. We can skip signing when we are not in one of the branches known for PPA upload. 

I can add that an reanable packaging.

**Comment by daschuer** [comment]
Done

**Comment by Be-ing** [comment]
```
dpkg-source: warning: source directory 'Mixxx-HEAD-r1-88741aa' is not <sourcepackage>-<upstreamversion> 'mixxx-2.3.0~beta'
dpkg-source: warning: .orig directory name Mixxx-HEAD-r1-88741aa.orig is not <package>-<upstreamversion> (wanted mixxx-2.3.0~beta.orig)
```
Do these warnings matter?

**Comment by daschuer** [comment]
This means we are not following the expected versioning schems. This also prevents us to move to version 3 source format.
This should be fixed after your versioning PR is merged.

**Comment by Be-ing** [comment]
Okay, LGTM then. @Holzhaus?

**Comment by daschuer** [comment]
I have just kept the current Sharma from Scons to do not break the version cintinoum on Launchpad. We schould do this only once.

**Review by Holzhaus — APPROVED** [review]
Thank you.

**Comment by daschuer** [comment]
All Green At: https://launchpad.net/~mixxx/+archive/ubuntu/mixxxbetas/+packages

**Comment by Be-ing** [comment]
I updated the main branch PPA Jenkins job too.

**Comment by daschuer** [comment]
Which Job exactly? 
I have just updated: 
https://builds.renegadetech.mixxx.org/job/master-nightly-ppa
for upload to: 
https://launchpad.net/~mixxx/+archive/ubuntu/nightlies/+packages

**Comment by Be-ing** [comment]
https://builds.renegadetech.mixxx.org/job/main-nightly-ppa/

Did you change the old job to use the main branch? I thought I disabled that job when the main branch was created. This is confusing. Now we have 2 Jenkins jobs building the same branch and uploading to the PPA.

**Comment by daschuer** [comment]
> I  updated the main branch PPA Jenkins job too.

Oh no! You have uploaded the 2.4 mixxx to the 2.3 ppa. 
Now every beta user will get the alpha version.

**Comment by Be-ing** [comment]
Whoops, I didn't see that I needed to change the PPA name. Can we delete those packages from the mixxxbetas PPA?

**Comment by daschuer** [comment]
I am just trying.

**Comment by daschuer** [comment]
> The easiest alternative to replace a broken source is always to upload a package with a higher version number and let the system automatically supersede and remove the older version. You should not attempt to use deletion requests to re-upload the same source version with different contents, as this is still prevented even after the content has been deleted.

**Comment by daschuer** [comment]
I am afraid Launchpad will reject packages with a smaller version number. We have now a bad situation. 
I will try to upload a newer 2.3 version. If this fails. I am trying to delete and restore the PPA. 

> When you no longer need your PPA, you can delete it. This deletes all of the PPA's packages, and removes the repository from ppa.launchpad.net. You'll have to wait up to an hour before you can recreate a PPA with the same name.

**Comment by Be-ing** [comment]
:( sorry

**Comment by daschuer** [comment]
It seams to be repaired. Puh ... 
I wonder how many users are affected from a undesired alpha update.

**Comment by Be-ing** [comment]
In 3 hours? Probably not many.

**Comment by Be-ing** [comment]
Thank you for taking care of that quickly.

**Comment by daschuer** [comment]
Download Stats: 

4 x  2.4.0~alpha~pre~git7676-0ubuntu1~main~git7676~bionic
21 x 2.4.0~alpha~pre~git7676-0ubuntu1~main~git7676~focal
3 x 2.4.0~alpha~pre~git7676-0ubuntu1~main~git7676~groovy

I think we have an issue. How can we reach theses people.

**Comment by Be-ing** [comment]
I don't think we have any way to track who that is (nor would I want that). I think it would be a good idea to post on the forum about this and instructions for how to fix it (uninstall then reinstall the package?).

**Comment by daschuer** [comment]
Yes. the correct betas are up now. So it should work. 
I am afraid that these people will only notice by the chance that they have been switch to alpha. 
Unfortunately, without the manual removal they will never get an update.
